src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,173 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private static final String INVALID_INDEX_ERROR_MESSAGE = "Invalid index";
+
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    public MyLinkedList() {
+
+    }
+
     @Override
     public void add(T value) {
+        if (isEmpty()) {
+            addFirstNode(value);
+        } else {
+            addLastNode(value);
+        }
+        ++size;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException(INVALID_INDEX_ERROR_MESSAGE);
+        }
+        if (isEmpty()) {
+            addFirstNode(value);
+        } else if (index == size) {
+            addLastNode(value);
+        } else {
+            Node<T> destNode = getNode(index);
+            link(destNode, value);
+        }
+        ++size;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T t : list) {
+            add(t);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException(INVALID_INDEX_ERROR_MESSAGE);
+        }
+        return getNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException(INVALID_INDEX_ERROR_MESSAGE);
+        }
+        T pastValue = getNode(index).value;
+        getNode(index).value = value;
+        return pastValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException(INVALID_INDEX_ERROR_MESSAGE);
+        }
+        Node<T> node = getNode(index);
+        unlink(node);
+        --size;
+        return node.value;
     }
 
     @Override
     public boolean remove(T object) {
+        for (int i = 0; i < size; i++) {
```

Comment:
Don't forget that you can have `null` as a method parameter in the `remove(T object)` method. Ensure that the logic correctly handles `null` values.